File: apps/etterlatte-saksbehandling-ui/client/src/shared/api/trygdetid.ts
import { apiClient, ApiResponse } from '~shared/api/apiClient'
import { JaNei } from '~shared/types/ISvar'

export const hentTrygdetider = async (behandlingId: string): Promise<ApiResponse<ITrygdetid[]>> =>
  apiClient.get<ITrygdetid[]>(`/trygdetid_v2/${behandlingId}`)

export const opprettTrygdetider = async (behandlingId: string): Promise<ApiResponse<ITrygdetid[]>> =>
  apiClient.post(`/trygdetid_v2/${behandlingId}`, {})

export const overstyrTrygdetid = async (overstyring: ITrygdetidOverstyring): Promise<ApiResponse<ITrygdetid>> =>
  apiClient.post(`/trygdetid_v2/${overstyring.behandlingId}/overstyr`, { ...overstyring })

export const setTrygdetidYrkesskade = async (yrkesskade: ITrygdetidYrkesskade): Promise<ApiResponse<ITrygdetid>> =>
  apiClient.post(`/trygdetid_v2/${yrkesskade.behandlingId}/yrkesskade`, { ...yrkesskade })

export const lagreTrygdetidgrunnlag = async (args: {
  behandlingId: string
  trygdetidId: string
  trygdetidgrunnlag: OppdaterTrygdetidGrunnlag
}): Promise<ApiResponse<ITrygdetid>> =>
  apiClient.post(`/trygdetid_v2/${args.behandlingId}/${args.trygdetidId}/grunnlag`, {
    ...args.trygdetidgrunnlag,
  })

export const slettTrygdetidsgrunnlag = async (args: {
  behandlingId: string
  trygdetidId: string
  trygdetidGrunnlagId: string
}): Promise<ApiResponse<ITrygdetid>> =>
  apiClient.delete<ITrygdetid>(
    `/trygdetid_v2/${args.behandlingId}/${args.trygdetidId}/grunnlag/${args.trygdetidGrunnlagId}`
  )

export const oppdaterStatus = async (behandlingId: string): Promise<ApiResponse<StatusOppdatert>> =>
  apiClient.post(`/trygdetid_v2/${behandlingId}/oppdater-status`, {})

export const oppdaterOpplysningsgrunnlag = async (behandlingId: string): Promise<ApiResponse<ITrygdetid>> =>
  apiClient.post(`/trygdetid_v2/${behandlingId}/oppdater-opplysningsgrunnlag`, {})

export interface StatusOppdatert {
  statusOppdatert: boolean
}

export interface TrygdetidAvtaleOptions {
  kode: string
  beskrivelse: string
}

export interface TrygdetidAvtaleDato extends TrygdetidAvtaleOptions {
  fraDato: Date
}

export interface TrygdetidAvtale extends TrygdetidAvtaleOptions {
  kode: string
  beskrivelse: string
  fraDato: Date
  datoer: TrygdetidAvtaleDato[]
}

export interface TrygdetidAvtaleKriteria extends TrygdetidAvtaleOptions {
  kode: string
  beskrivelse: string
  fraDato: Date
}

export interface Trygdeavtale {
  id: string
  behandlingId: string
  avtaleKode: string
  avtaleDatoKode?: string
  avtaleKriteriaKode?: string
  personKrets?: JaNei
  arbInntekt1G?: JaNei
  arbInntekt1GKommentar?: string
  beregArt50?: JaNei
  beregArt50Kommentar?: string
  nordiskTrygdeAvtale?: JaNei
  nordiskTrygdeAvtaleKommentar?: string
  kilde: {
    tidspunkt: string
    ident: string
  }
}

export interface TrygdeavtaleRequest {
  id?: string
  avtaleKode: string
  avtaleDatoKode?: string
  avtaleKriteriaKode?: string
  arbInntekt1G?: JaNei
  arbInntekt1GKommentar?: string
  beregArt50?: JaNei
  beregArt50Kommentar?: string
  nordiskTrygdeAvtale?: JaNei
  nordiskTrygdeAvtaleKommentar?: string
}

export const hentAlleTrygdetidAvtaler = async (): Promise<ApiResponse<TrygdetidAvtale[]>> =>
  apiClient.get<TrygdetidAvtale[]>('/trygdetid/avtaler')

export const hentAlleTrygdetidAvtaleKriterier = async (): Promise<ApiResponse<TrygdetidAvtaleKriteria[]>> =>
  apiClient.get<TrygdetidAvtaleKriteria[]>('/trygdetid/avtaler/kriteria')

export const hentTrygdeavtaleForBehandling = async (args: {
  behandlingId: string
}): Promise<ApiResponse<Trygdeavtale>> => apiClient.get<Trygdeavtale>(`/trygdetid/avtaler/${args.behandlingId}`)

export const lagreTrygdeavtaleForBehandling = async (args: {
  behandlingId: string
  avtaleRequest: TrygdeavtaleRequest
}): Promise<ApiResponse<Trygdeavtale>> =>
  apiClient.post<Trygdeavtale>(`/trygdetid/avtaler/${args.behandlingId}`, { ...args.avtaleRequest })

export const opprettTrygdetidOverstyrtMigrering = async (args: {
  behandlingId: string
  overskriv?: boolean
}): Promise<ApiResponse<void>> =>
  apiClient.post(`/trygdetid_v2/${args.behandlingId}/migrering/manuell/opprett?overskriv=${args.overskriv}`, {})

export const oppdaterTrygdetidOverstyrtMigrering = async (args: {
  behandlingId: string
  trygdetidId: string
  anvendtTrygdetid: number
  prorataBroek?: IProrataBroek
}): Promise<ApiResponse<ITrygdetid>> =>
  apiClient.post(`/trygdetid_v2/${args.behandlingId}/migrering/${args.trygdetidId}/manuell/lagre`, {
    samletTrygdetidNorge: args.prorataBroek ? undefined : args.anvendtTrygdetid,
    samletTrygdetidTeoretisk: args.prorataBroek ? args.anvendtTrygdetid : undefined,
    prorataBroek: args.prorataBroek,
    overstyrt: true,
    yrkesskade: false,
  })

export interface ITrygdetid {
  id: string
  ident: string
  behandlingId: string
  beregnetTrygdetid?: IDetaljertBeregnetTrygdetid
  trygdetidGrunnlag: ITrygdetidGrunnlag[]
  opplysninger: IGrunnlagOpplysninger
  overstyrtNorskPoengaar: number | undefined
  opplysningerDifferanse: IOpplysningerDifferanse | undefined
}

export interface ITrygdetidOverstyring {
  id: string
  behandlingId: string
  overstyrtNorskPoengaar: number | undefined
}

export interface ITrygdetidYrkesskade {
  id: string
  behandlingId: string
  yrkesskade: boolean
}

export interface IGrunnlagOpplysninger {
  avdoedDoedsdato: IOpplysningsgrunnlag
  avdoedFoedselsdato: IOpplysningsgrunnlag
  avdoedFylteSeksten: IOpplysningsgrunnlag
  avdoedFyllerSeksti: IOpplysningsgrunnlag
}

export type IOpplysningsgrunnlag = {
  opplysning: string
  kilde: {
    type: string
    tidspunkt: string
  }
}

export interface IDetaljertBeregnetTrygdetid {
  resultat: IDetaljertBeregnetTrygdetidResultat
  tidspunkt: Date
}

export interface IDetaljertBeregnetTrygdetidResultat {
  faktiskTrygdetidNorge: IFaktiskTrygdetid
  faktiskTrygdetidTeoretisk: IFaktiskTrygdetid
  fremtidigTrygdetidNorge: IFremtidigTrygdetid
  fremtidigTrygdetidTeoretisk: IFremtidigTrygdetid
  samletTrygdetidNorge?: number
  samletTrygdetidTeoretisk?: number
  prorataBroek?: IProrataBroek
  overstyrt: boolean
  yrkesskade: boolean
  beregnetSamletTrygdetidNorge?: number
}

export interface IFaktiskTrygdetid {
  periode?: string
  antallMaaneder?: number
}

export interface IFremtidigTrygdetid {
  periode?: string
  antallMaaneder?: number
  opptjeningstidIMaaneder?: number
  mindreEnnFireFemtedelerAvOpptjeningstiden?: Boolean
  nordiskKonvensjon?: Boolean
}

export interface IProrataBroek {
  teller: number
  nevner: number
}

export interface ITrygdetidGrunnlag {
  id: string
  type: ITrygdetidGrunnlagType
  bosted: string
  periodeFra: string
  periodeTil: string
  beregnet?: IBeregnetTrygdetidGrunnlag
  kilde: {
    tidspunkt: string
    ident: string
  }
  begrunnelse?: string
  poengInnAar?: boolean
  poengUtAar?: boolean
  prorata?: boolean
}

export interface OppdaterTrygdetidGrunnlag {
  id?: string
  type: ITrygdetidGrunnlagType
  bosted: string
  periodeFra?: string
  periodeTil?: string
  begrunnelse?: string
  poengInnAar?: boolean
  poengUtAar?: boolean
  prorata?: boolean
}

export interface IBeregnetTrygdetidGrunnlag {
  dager: number
  maaneder: number
  aar: number
}

export enum ITrygdetidGrunnlagType {
  FAKTISK = 'FAKTISK',
  FREMTIDIG = 'FREMTIDIG',
}

export interface IOpplysningerDifferanse {
  differanse: Boolean
  oppdaterteGrunnlagsopplysninger: IGrunnlagOpplysninger
}

File: libs/etterlatte-trygdetid-model/src/main/kotlin/Trygdetid.kt
package no.nav.etterlatte.libs.common.trygdetid

import com.fasterxml.jackson.databind.JsonNode
import no.nav.etterlatte.libs.common.IntBroek
import no.nav.etterlatte.libs.common.tidspunkt.Tidspunkt
import java.time.LocalDate
import java.time.Period
import java.util.UUID

data class TrygdetidDto(
    val id: UUID,
    val ident: String,
    val behandlingId: UUID,
    val beregnetTrygdetid: DetaljertBeregnetTrygdetidDto?,
    val trygdetidGrunnlag: List<TrygdetidGrunnlagDto>,
    val opplysninger: GrunnlagOpplysningerDto,
    val overstyrtNorskPoengaar: Int?,
    val opplysningerDifferanse: OpplysningerDifferanse,
    val begrunnelse: String?,
)

const val UKJENT_AVDOED = "UKJENT_AVDOED"

data class TrygdetidOverstyringDto(
    val id: UUID,
    val behandlingId: UUID,
    val overstyrtNorskPoengaar: Int?,
)

data class TrygdetidYrkesskadeDto(
    val id: UUID,
    val behandlingId: UUID,
    val yrkesskade: Boolean,
)

data class TrygdetidBegrunnelseDto(
    val id: UUID,
    val behandlingId: UUID,
    val begrunnelse: String?,
)

data class GrunnlagOpplysningerDto(
    val avdoedDoedsdato: OpplysningsgrunnlagDto?,
    val avdoedFoedselsdato: OpplysningsgrunnlagDto?,
    val avdoedFylteSeksten: OpplysningsgrunnlagDto?,
    val avdoedFyllerSeksti: OpplysningsgrunnlagDto?,
) {
    companion object {
        fun tomt() =
            GrunnlagOpplysningerDto(
                null,
                null,
                null,
                null,
            )
    }
}

data class OpplysningsgrunnlagDto(
    val opplysning: JsonNode,
    val kilde: OpplysningkildeDto,
)

data class OpplysningkildeDto(
    val type: String,
    val tidspunkt: String,
)

data class DetaljertBeregnetTrygdetidDto(
    val resultat: DetaljertBeregnetTrygdetidResultat,
    val tidspunkt: Tidspunkt,
)

data class TrygdetidGrunnlagDto(
    val id: UUID?,
    val type: String,
    val bosted: String,
    val periodeFra: LocalDate,
    val periodeTil: LocalDate,
    val kilde: TrygdetidGrunnlagKildeDto?,
    val beregnet: BeregnetTrygdetidGrunnlagDto?,
    val begrunnelse: String?,
    val poengInnAar: Boolean,
    val poengUtAar: Boolean,
    val prorata: Boolean,
)

data class TrygdetidGrunnlagKildeDto(
    val tidspunkt: String,
    val ident: String,
)

data class BeregnetTrygdetidGrunnlagDto(
    val dager: Int,
    val maaneder: Int,
    val aar: Int,
)

data class MigreringOverstyringDto(
    val ident: String,
    val detaljertBeregnetTrygdetidResultat: DetaljertBeregnetTrygdetidResultat,
)

data class StatusOppdatertDto(
    val statusOppdatert: Boolean,
)

data class DetaljertBeregnetTrygdetidResultat(
    val faktiskTrygdetidNorge: FaktiskTrygdetid?,
    val faktiskTrygdetidTeoretisk: FaktiskTrygdetid?,
    val fremtidigTrygdetidNorge: FremtidigTrygdetid?,
    val fremtidigTrygdetidTeoretisk: FremtidigTrygdetid?,
    val samletTrygdetidNorge: Int?,
    val samletTrygdetidTeoretisk: Int?,
    val prorataBroek: IntBroek?,
    val overstyrt: Boolean,
    val yrkesskade: Boolean,
    val beregnetSamletTrygdetidNorge: Int?,
    val overstyrtBegrunnelse: String? = null,
) {
    companion object {
        fun fraSamletTrygdetidProrata(
            anvendtTrygdetid: Int,
            prorataBroek: IntBroek?,
        ) = DetaljertBeregnetTrygdetidResultat(
            faktiskTrygdetidNorge = null,
            faktiskTrygdetidTeoretisk = null,
            fremtidigTrygdetidNorge = null,
            fremtidigTrygdetidTeoretisk = null,
            samletTrygdetidNorge = null,
            samletTrygdetidTeoretisk = anvendtTrygdetid,
            prorataBroek = prorataBroek,
            overstyrt = true,
            yrkesskade = false,
            beregnetSamletTrygdetidNorge = null,
        )
    }
}

data class FaktiskTrygdetid(
    val periode: Period,
    val antallMaaneder: Long,
) {
    fun avrundet(): FaktiskTrygdetid {
        val avrundetPeriode = this.periode.justertForDager()
        return this.copy(periode = avrundetPeriode, antallMaaneder = avrundetPeriode.toTotalMonths())
    }
}

fun Period.justertForDager(): Period {
    val antallMaaneder =
        if (this.normalized().days > 0) {
            this.toTotalMonths() + 1L
        } else {
            this.toTotalMonths()
        }
    return Period.ofMonths(antallMaaneder.toInt()).normalized()
}

data class FremtidigTrygdetid(
    val periode: Period,
    val antallMaaneder: Long,
    val opptjeningstidIMaaneder: Long,
    val mindreEnnFireFemtedelerAvOpptjeningstiden: Boolean,
) {
    fun avrundet(): FremtidigTrygdetid {
        val avrundetPeriode = this.periode.justertForDager()
        return this.copy(periode = avrundetPeriode, antallMaaneder = avrundetPeriode.toTotalMonths())
    }
}

data class OpplysningerDifferanse(
    val differanse: Boolean,
    val oppdaterteGrunnlagsopplysninger: GrunnlagOpplysningerDto,
)
